### ISSUE TITLE
User ranks query class

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/wordpoints#684]
As pointed out in https://github.com/WordPoints/wordpoints/issues/642#issuecomment-308842625, we perform a basic user ranks query in `wordpoints_get_users_with_rank()`. However, because it is contained in a procedural function, it is more difficult to utilize to get the results in batches, instead of pulling them all out at once. It just isn't designed for that purpose. We've generally been moving toward query classes for this, similar to `WP_Query` and kin found in WordPress. And it seems that is what is needed here.

The main problem we'll have to deal with is #682. It would be possible for us to build a query class for user ranks without fixing that, however, it would severely limit its features. Because we'd have to join on the users table, we would basically be limited to querying by rank ID and user ID. Ordering and more complex searches would be pretty much impossible.

In other words, because the table doesn't contain all of the data (some of it is implicit and must be inferred), it wouldn't be able to be a generic query class for the table data. It would have to be more of just a "get users with rank" query class (the type of query that that function performs). That's the main thing that we need it for right now anyway though.

Basically, it would only support querying by a single rank ID.

**Comment by JDGrimes** [comment]
It seems almost like it would be more work to trim down a child of our db query class than it would be to fix the issue here. Though I guess we could just only document the `rank_id` query arg as being supported, and not actually actively rip out or warn about other query args if supplied.

In reality though, I guess the query *could* be used with other args involving other ranks, just not any rank that isn't the base rank in a group. Or any time that handling for base ranks isn't important. So maybe it is best not to suppress the other query args anyway, just to be sure to document the query's shortcomings in regard to base ranks.

**Comment by JDGrimes** [comment]
But what about the sub queries? Because of them using limits will be more complex, because we have to use ordering with the limit for consistency in results, which I think will be problematic. And it will also be less performant, because the entire sub query likely has to be run *before* the limit gets applied.

So although batching is an improvement needed for the sake of memory, if we have to build the query in this manner we'll be introducing other performance issues at the database level.

It might be doable, but it's clear that it isn't going to be pretty.

**Comment by JDGrimes** [comment]
As far as the limits go, I think that our assessment is correct. The `UNION ALL` would pull all results out, and then apply the outer limit on the temporary table containing all results that it had created internally.

And we can't use the limits on the individual queries, because the pagination won't work correctly (each query will return the full limit).

I had thought, however, that we could try and place a limit of half of the outer limit on each of the queries. So our outer limit could be 1000, and each query would be paged by 500. Once one of the queries ran out of results, we'd get only 500 results back, of course, instead of the full 1000, but it would work. In our case it is not so important that we get a particular amount.

Another option is to [use a condition on one of the column values to limit the range of results returned](https://stackoverflow.com/a/10213865/1924128). We could use the user ID, for example. However, if there was a large stretch of users that had been deleted, we'd end up returning very few or no results from some of the "pages".

**Comment by JDGrimes** [comment]
Maybe in the end though, the least expensive thing to do would be to just fill in the base rank users on demand when we are making a query involving the base rank. Then we can avoid the union/join entirely.

It wouldn't just be a one-time thing though, because ranks aren't assigned to new users as they are added. But we could change that. 

This would make it so that the ranks wouldn't be filled until a second rank was added to the rank group.

**Comment by JDGrimes** [comment]
By making it a one-time thing, we could keep track of a list of base ranks for which it had already been triggered. Then we'd avoid doing it more than once. I was thinking that we could just keep track of this per rank group, but then I considered the possibility of edge cases when a base rank is moved/removed/replaced. So we could just do it by rank ID. However, I have a feeling that there would be other edge cases relating to that anyway. Although technically possible, messing with base ranks in that way is really not a good idea.

And I guess really, either way we will have edge cases, whether we do it by rank ID or rank group. But the ID is smaller, and perhaps has fewer edge cases, so let's go with that.

**Comment by JDGrimes** [comment]
I suppose we should make the method that fills in the ranks public, so that if someone needs to perform some other kind of query that may involve a base rank, they could fill the info in first. (We can't do this automatically, because we may at times have no way of knowing whether a base rank is relevant or not.)
